<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Tutorial</title>

<script src="tutorial_files/header-attrs-2.26/header-attrs.js"></script>
<script src="tutorial_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="tutorial_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="tutorial_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="tutorial_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="tutorial_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="tutorial_files/navigation-1.1/tabsets.js"></script>
<link href="tutorial_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="tutorial_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="custom.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Tutorial</h1>

</div>


<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Welcome to <code>RiboCrypt</code></p>
<p><code>RiboCrypt</code> is an R package and web server for interactive
visualization and analysis in genomics. <code>RiboCrypt</code> works
with most NGS-based methods, but much emphasis is put on Ribo-seq data
visualization. RiboCrypt uses ORFik experiment objects, so to ensure
compatibility use
<a href="https://doi.org/10.1186/s12859-021-04254-w" target="_blank">ORFik</a>
and
<a href="https://github.com/rc-biotech/massiveNGSpipe" target="_blank">massiveNGSpipe</a>
for processing own (unpublished) data. If you encounter any issues,
please, contact us using the info in the footnote.</p>
<p>The following is web server walk through listing all utilities and
options therein.</p>
<div id="ribo-seq-overview" class="section level2">
<h2>Ribo-seq overview</h2>
<p>For more general material about ribosome profiling, we suggest
reading
<a href="https://cshperspectives.cshlp.org/content/11/5/a032698.full.pdf+html" target="_blank">this
review</a>. For more technical explanation of data processing steps and
specific analyses, it’s best to see
<a href="https://bioconductor.org/packages/release/bioc/html/ORFik.html" target="_blank">ORFik</a>
Overview vignette, especially chapter 6
<a href="https://bioconductor.org/packages/release/bioc/vignettes/ORFik/inst/doc/ORFikOverview.html#riboseq-footprints-automatic-shift-detection-and-shifting" target="_blank">“RiboSeq
footprints automatic shift detection and shifting”</a>.</p>
</div>
</div>
<div id="browser" class="section level1">
<h1>Browser</h1>
<p>The browser is the basic coverage display page. It contains a
selection panel on the left side and a display window (browser) on the
right. It displays coverage of sequencing data in either transcript
coordinates (default, collapsed introns), or genomic coordinates. The
following (Fig. 1), SRD5A1 gene,
<a href="https://doi.org/10.21203/rs.3.rs-5390104/v1" target="_blank">has
recently been shown to be decoded in three frames</a>.</p>
<p><img src="../images/tutorial_fig1.png" width="1000px" height="600px" style="display: block; margin: auto auto auto 0;" /></p>
<p>Figure 1. SRD5A1 gene displayed using default options. Using <a
href="#API">API</a>, this browser view can be re-generated using this <a
href="https://ribocrypt.org/?&amp;dff=all_merged-Homo_sapiens_modalities&amp;gene=SRD5A1-ENSG00000145545&amp;tx=ENST00000274192&amp;library=RFP&amp;frames_type=area&amp;kmer=9&amp;log_scale=FALSE&amp;extendLeaders=0&amp;extendTrailers=0&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">link</a>.</p>
<div id="browser-window" class="section level2">
<h2>Browser window</h2>
<p>When you press the “plot” button on the selection panel, the selected
data will be displayed according to specified options, explained later
in the tutorial.</p>
<p>The browser window consists of the specific parts:</p>
<ol style="list-style-type: decimal">
<li>Single or multiple density tracks are displayed on top. By default
Ribo-seq is rendered in 3 colors, where</li>
</ol>
<ul>
<li>red is 0 frame, the start frame of reference transcript.</li>
<li>green is +1 frame</li>
<li>blue is +2 frame</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Sequence track (top middle), displayes DNA sequence when zoomed in
(&lt; 100nt)</li>
<li>Annotation track (middle), the annotation track displays the
transcript annotation, together with black bars that is displayed on top
of the data track. CDS and other annotations are color-coded according
to relative reading frame</li>
<li>Frame track (bottom), the 3 frames displayed with given color
bars:</li>
</ol>
<ul>
<li><p>white (start codon, ATG)</p></li>
<li><p>black (stop codons, TAA | TAG | TGA)</p></li>
<li><p>brown (custom motifs, e.g. NTG all start codon alternatives)</p>
<p>When zoomed in, the amino acid sequence is displayed within each
frame.</p></li>
</ul>
<ol start="5" style="list-style-type: decimal">
<li>Additional displays, for example protein structure viewer, which
appears after clicking on isoform ID on annotation track.</li>
</ol>
<p>In the example above (Fig. 1), notice green coverage preceding CDS
(in Blue). It corresponds to overlapping open reading frame, so it is
clear that there are two regions that undergo translation simultaneously
in this locus.</p>
</div>
<div id="selection-panel-browser" class="section level2">
<h2>Selection panel (browser)</h2>
<p>The display panel shows the primary settings, (study, gene, sample,
etc):</p>
<div id="exp" class="section level3">
<h3>Experiment</h3>
<ul>
<li>Select an organism: Either select “ALL” to keep all experiments, or
select a specific organism to select display only that subset of
experiments in experiment select tab.</li>
<li>Select an experiment: The experiments contain study names combined
with organism (some studies are multi species, so sometimes one study
have multiple experiments). Select which one you want. There also exist
merged experiments (all samples merged for the organism, etc). RNA-seq
experiments have an addition “_RNA” as suffix.</li>
</ul>
</div>
<div id="gene" class="section level3">
<h3>Gene</h3>
<ul>
<li>Select a gene: A gene can be selected using:
<ul>
<li>Gene id (ENSEMBL)</li>
<li>Gene symbol (HGNC, etc)</li>
</ul></li>
<li>Select a transcript: A transcript isoform of the given gene above,
default is Ensembl canonical isoform, other isoforms can be
selected.</li>
</ul>
</div>
<div id="lib" class="section level3">
<h3>Library</h3>
<p>Usually each experiment have multiple libraries. In the case of
merged experiments, different modalities (RFP, RNA, disome, TI, see Fig.
2) can be selected in this field. Select which one to display, if you
select multiple libraries they will be stacked in the browser as
multiple tracks.</p>
<p>Library are by default named:</p>
<ul>
<li>Library type (RFP, RNA etc),</li>
<li>Condition (WT, KO (wild type, knock out ) etc)</li>
<li>Stage/timepoint (5h, 1d (5 hours, 1 day) etc)</li>
<li>fraction (chx, cytosolic, ATF4 (ribosomal inhibitor, cell fraction,
gene) etc)</li>
<li>replicate (technical/biological replicate number (r1, r2, r3))</li>
</ul>
<p>The resuting name could for example be:</p>
<ul>
<li>RFP_WT_5h_chx_cytosolic_r1</li>
</ul>
<p>It’s normal to see that if condition is KO (knockout), the fraction
column contains a gene name (the name of the gene that was knocked out).
Currently the best way to find SRR run number for respective sample is
to go to metadata tab and search for the study.</p>
</div>
<div id="linetype" class="section level3">
<h3>Display type</h3>
<ul>
<li>Select reading frames display type:
<ul>
<li>lines (single line per frame, most clear for middle distance (&gt;
100 nt))</li>
<li>columns (single bars per position, most clear for single nucleotide
resolution)</li>
<li>stacks (Area under curve, stacked, most clear for large regions
(&gt; 1000 nt))</li>
<li>area (Area under curve, with semitransparent overlapping frames,
most clear for large regions (&gt; 1000 nt))</li>
<li>heatmap (good for displaying large number of libraries, see
metabrowser for display of thousands of libraries)</li>
</ul></li>
<li>K-mer length: sliding window of selected length applying averaging
coverage per frame. K-mer = 1 means unaltered data. When looking at a
large region (&gt; 100nt), pure coverage can usually be hard to inspect
due to intrinsic Ribo-seq spikyness. Using K-mer length &gt; 1 (9, the
default, is a good starting point), you can easily look at patterns over
larger regions.</li>
</ul>
<p>Notice, how in figure 2 reducing K-mer length to 1, changing display
type to columns and adding TI-seq (Translation Initiation profiling,
samples treated with translation initiation inhibitors - either
lactimidomycin or harringtonin) enables detection of translation in all
three reading frames. In fact, experienced user can notice drop of
signal in red frame after stop codon of the corresponding ORF (hidden
behind most abundant, green), further corroborating overlapping
translation. If this process seems somewhat imprecise and not fully
defined - it is by design - annotating such elusive phenomena is
notoriously difficult to tackle algorithmically, and doing so
successfully requires deep intuition in how the signal and noise behave,
so manual investigation into many instances like this. Thus, RiboCrypt
can be treated as a hypothesis building tool, rather a database of fixed
answers. <a
href="https://ribocrypt.org/?&amp;dff=all_merged-Homo_sapiens_modalities&amp;gene=SRD5A1-ENSG00000145545&amp;tx=ENST00000274192&amp;library=RFP,TI&amp;frames_type=columns&amp;kmer=1&amp;log_scale=FALSE&amp;extendLeaders=0&amp;extendTrailers=0&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">Try</a>
modifying yourself, for example switching on RNA-seq track, or altering
smoothing window width.</p>
<p><img src="../images/tutorial_fig2.png" width="500px" height="600px" style="display: block; margin: auto auto auto 0;" /></p>
<p>Figure 2. Comparison of display types and smoothing window
widths.</p>
</div>
</div>
<div id="settings" class="section level2">
<h2>Selection panel (Settings)</h2>
<p>Settings tab contains additional parameters:</p>
<ul>
<li><p>5’ extension (extend viewed window upstream)</p></li>
<li><p>3’ extension (extend viewed window downstream)</p></li>
<li><p>Custom sequences highlight (Motif search, given in brown color,
support IUPAC, examples: CTG or NTG)</p></li>
<li><p>Genomic region (Browse genomic window instead of gene, syntax:
chromosome:start-stop:strand, human/mouse/zebrafish: 1:10000-20000:+ ,
yeast: I:10000-20000:+. Both 1 and chr1 works, conversion will be done
automatically)</p></li>
<li><p>Zoom interval (start with a zoom and highlight on a subsection):
Either specified in tx/view coordinates, i.e. 20:50 will give zoom on
region (20-10):(50+10) = 10:60, and a highlight color (light yellow) at
coordinates 20:50. You can also specify in genomic coordinates as above
for genomic region (remember the genomic coordinates must then be within
the gene / region you are displaying).</p></li>
<li><p>Genomic View (Activate/deactivate genomic view, displaying full
introns, see Figure 3)</p></li>
<li><p>Full annotation (display all transcript isoforms annotation or
just the selected isoform)</p></li>
<li><p>uORF annotation (display candidate uORFs in the annotation track)
this is not the predicted set, but all possible uORFs by seqnece. For
predicted set, check the predicted translons checkbox.</p></li>
<li><p>Predicted translons - display <a href="#translons">predicted
translons</a> in the annotation track. Try turning it on for the <a
href="https://ribocrypt.org/?&amp;dff=all_merged-Homo_sapiens_modalities&amp;gene=SRD5A1-ENSG00000145545&amp;tx=ENST00000274192&amp;library=RFP&amp;frames_type=area&amp;kmer=9&amp;log_scale=FALSE&amp;extendLeaders=0&amp;extendTrailers=0&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">example</a>
in Figures 1 and 2 - notice that the red (second) translon was
missed</p></li>
<li><p>Log scale (Log scale the coverage, reduces effect of extreme
peaks)</p></li>
<li><p>Protein structures (If you click the annotation name of a
transcript in the plot panel it will display the alpha-fold protein
colored by the ribo-seq data displayed in the plot panel)</p></li>
<li><p>PhyloP (PhyloP conservation track will be added on the
bottom)</p></li>
<li><p>Split color frame (Split riboseq signal into colors by frame)
i.e. red, green and blue. If this is on, then Ribo-seq and Ti-seq will
be displayed with colors split by frame.</p></li>
<li><p>Frames subset (If split color frame is TRUE, you can display only
wanted color)</p></li>
<li><p>Summary top track (Add an additional plot track on top,
summarizing all selected libraries)</p></li>
<li><p>Select Summary display type (same as frames display type above,
but for the summary track)</p></li>
<li><p>Export format (Choose plot export format from plotly controls.
The default is svg vector graphics that allows for high
customizability)</p></li>
<li><p>Plot (red button, render selected gene/region as plotly html,
with specified settings)</p></li>
</ul>
<p>Uniquely, RiboCrypt allows for toggling between transcriptomic and
genomic views, while displaying coding exons in colors according to the
correct reading frame. Moreover, extensions allow for exploration of
large chunks of the genome, even tens of thousands of bases. <a
href="https://ribocrypt.org/?&amp;dff=all_merged-Saccharomyces_cerevisiae_modalities&amp;gene=EFM5-YGR001C&amp;tx=YGR001C_mRNA&amp;library=RFP&amp;frames_type=area&amp;kmer=5&amp;log_scale=FALSE&amp;extendLeaders=15000&amp;extendTrailers=15000&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">Try</a>
finding all genes in this browser window, and then turn full annotation
on to see what you missed! (loading may be a bit slow, but displaying
over region 30 thousands bases long here isn’t practical)</p>
<p><img src="../images/tutorial_fig3.png" width="900px" height="600px" style="display: block; margin: auto auto auto 0;" /></p>
<p>Figure 3. Comparison between transcriptomic and genomic view on yeast
<a
href="https://ribocrypt.org/?&amp;dff=all_merged-Saccharomyces_cerevisiae_modalities&amp;gene=EFM5-YGR001C&amp;tx=YGR001C_mRNA&amp;library=RFP&amp;frames_type=area&amp;kmer=5&amp;log_scale=FALSE&amp;extendLeaders=1&amp;extendTrailers=150&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">EFM5</a>
and <a
href="https://ribocrypt.org/?&amp;dff=all_merged-Saccharomyces_cerevisiae_modalities&amp;gene=ABP140-YOR239W&amp;tx=YOR239W_mRNA&amp;library=RFP&amp;frames_type=area&amp;kmer=5&amp;log_scale=FALSE&amp;extendLeaders=1&amp;extendTrailers=150&amp;viewMode=FALSE&amp;other_tx=FALSE&amp;add_uorfs=FALSE&amp;genomic_region=&amp;customSequence=&amp;phyloP=FALSE&amp;summary_track=FALSE&amp;go=TRUE">ABP140</a>
genes.</p>
</div>
<div id="section" class="section level2">
<h2></h2>
</div>
</div>
<div id="mbrowser" class="section level1">
<h1>Mega Browser</h1>
<p>Display all samples for a specific organism over selected gene (Fig.
5). This tab does not use bigwig files to load (as that would be very
slow). It uses precomputed fst files of coverage over all libraries.
Note: Not all isoforms are computed, by default the longest isoform is
computed.</p>
<p><img src="../images/tutorial_fig5.png" width="1000px" height="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p>Figure 5. Heatmap of thousands of libraries coverage over ATF4
transcript. Clustered using k-means = 2, with summary track displayed on
top.</p>
<div id="selection-panel-browser-1" class="section level2">
<h2>Selection panel (Browser)</h2>
<p>Organism, experiment and gene explained above</p>
<ul>
<li>Group on: the metadata column to order plot by</li>
<li>K-means clusters: How many k-means clusters to use, if &gt; 1, Group
will be sorted within the clusters, but K-means have priority.</li>
</ul>
</div>
<div id="selection-panel-settings" class="section level2">
<h2>Selection panel (Settings)</h2>
<ul>
<li>Normalization (all scores are tpm normalized and log scaled)
<ul>
<li>transcriptNormalized (each sample counts sum to 1) (default)</li>
<li>Max normalized (each position count divided by maximum)</li>
<li>zscore (zscore normalization, (count * mean / sd) variance scaled
normalization)</li>
<li>tpm (raw tpm of counts, is very sensitive to extreme peaks)</li>
</ul></li>
<li>Color theme: Which color theme to use</li>
<li>Color scale multiplier: how much to amplify the color signal (if all
is single color, try to reduce or increase this depending on which color
is the majority)</li>
<li>Sort by other gene: Sort heatmap on expression of another gene
(increasing). A line plot will be shown left that display the expression
of that other gene per row. Also the enrichment analysis is done on bins
of the other genes expression instead of metadata term.</li>
<li>Summary top track (same as in browser, the aggregate of all rows,
useful to see the frame information, which is not represented in the
heatmap)</li>
<li>Split by frame: (Not active yet!): When completed will display the
heatmap with frame information, so rgb colors from white to dark red
etc.</li>
<li>Summary display type: Same as in browser</li>
<li>Region to view: By default uses entire transcript. Can subset to
only view cds or leader+cds etc.</li>
<li>Select a transcript: Which transcript isoforms not all isoforms have
computed fsts. By default the longest isoforms have been computed for
all genes.</li>
<li>K-mer length: Smoothens out the signal by applying a mean sliding
window, default 1 (off)</li>
</ul>
</div>
<div id="statistics-tab" class="section level2">
<h2>Statistics tab</h2>
<p>This tab gives the statistics of over representation analysis per
cluster of the metabrowser plot. Using chi squared test, it gives the
residuals per term from metadata (like tissue, cell-line etc). If a
value is bigger than +/- 3, it means it is quite certain this is over
represented. This is shown as a red line.</p>
<p>If no clustering was applied, this tab gives the number of items per
metadata term (40 brain samples, 30 kidney samples etc).</p>
</div>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<p>This mode is very intensive on CPU, so it requires certain
pre-computed results for the back end. That is namely: - Premade
collection experiments (an ORFik experiment of all experiments per
organism) - Premade collection count table and library sizes (for
normalizations purpose) - Premade fst serialized coverage calculation
per gene (for instant loading of coverage over thousands of
libraries)</p>
<p>Note that on the live app, the human collection (4000 Ribo-seq
samples) takes around 30 seconds to plot for a ~ 2K nucleotides gene,
~99% of the time is spent on rendering the plot, not actual computation.
Future investigation into optimization will be done.</p>
</div>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Here we collect the analysis possibilities, which are usually
operating on meta-gene or multi locus scale.</p>
<div id="heatmap" class="section level2">
<h2>Heatmap</h2>
<p>This tab displays a heatmap of coverage per readlength at a specific
region (like start site of coding sequences) over all genes
selected.</p>
<p><img src="../images/tutorial_fig4.png" width="1000px" height="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p>Figure 4. Metagene per-readlength heatmap of before (upper panel) and
after (bottom panel) P-shifting. Notice emerging periodicity.</p>
<div id="selection-panel-heatmap" class="section level3">
<h3>Selection panel (heatmap)</h3>
<p>Study and gene select works same as for browser specified above. In
addition to have the option to specify all genes (default).</p>
<ul>
<li>Select libraries. Currently Only 1 library can be selected in
heatmap mode.</li>
<li>View region Select one of:
<ul>
<li>Start codon</li>
<li>Stop codon</li>
</ul></li>
<li>Normalization Normalization mode for data display, select one of:
<ul>
<li>transcriptNormalized (each gene counts sum to 1)</li>
<li>zscore (zscore normalization, will give better overview if 1
readlength is extreme)</li>
<li>sum (raw sum of counts, is very sensitive to extreme peaks)</li>
<li>log10sum (log10 sum of counts, is less sensitive to extreme
peaks)</li>
</ul></li>
<li>Min Readlength The minimum readlength to display</li>
<li>Max readlength The maximum readlength to display</li>
</ul>
</div>
<div id="selection-panel-settings-1" class="section level3">
<h3>Selection panel (settings)</h3>
<p>Here additional options are shown:</p>
<ul>
<li><p>5’ extension (extend viewed window upstream from point, default
30)</p></li>
<li><p>3’ extension (extend viewed window downstreamfrom point, default
30)</p></li>
<li><p>Summary top track - summarized coverage from all
readlengths</p></li>
<li><p>p-shifted - display either P-sites (default), or reads 5’
ends.</p></li>
</ul>
<p>Metagene analysis with heatmap module can be used, for example, to
investigate how well P-site positioning was performed (Fig. 4)</p>
</div>
</div>
<div id="codon-analysis" class="section level2">
<h2>Codon analysis</h2>
<p>This tab displays a heatmap of codons dwell times over all genes
selected, for both A and P sites. When pressing “Differential” you swap
to a between library differential codon dwell time comparison (minimum 2
libraries selected is required for this method!)</p>
<div id="display-panel-codon" class="section level3">
<h3>Display panel (codon)</h3>
<p>Study and gene select works same as for browser specified above. In
addition to have the option to specify all genes (default).</p>
<ul>
<li>Select libraries (multiple allowed)</li>
</ul>
<div id="codonfilt" class="section level4">
<h4>Filters</h4>
<ul>
<li>Codon filter value (Minimum reads in ORF to be included)</li>
<li>Codon score, all scores are normalized for both codon and count per
gene level (except for sum):
<ul>
<li>percentage (percentage use relative to max codon, transcript
normalized percentages)</li>
<li>dispersion(NB) (negative binomial dispersion values)</li>
<li>alpha(DMN) (Dirichlet-multinomial distribution alpha parameter)</li>
<li>sum (raw sum, (a very biased estimator, since some codons are used
much more than others!))</li>
</ul></li>
</ul>
</div>
<div id="codon-plot" class="section level4">
<h4>Codon plot</h4>
<p>Display is the score per codon (amino acid), in addition there are 2
custom “amino acids”, * as in * : TGA, means TGA is a stop codon (last
codon in CDS). Similar is #, as in # : ATG which means ATG as start
codon (first codon in CDS). For P-sites start codons should be enriched,
while for A-sites there should be a richer variability, often with a
small enrichment for stop-codons. We will implement a richer model
eventually using the more correct negative binomial relationships
between E, P and A sites, i.g. the motif PPP (triplet-proline in E,P,A
site) is much stronger than a single P in the A site etc. alone etc.</p>
</div>
</div>
</div>
<div id="section-1" class="section level2">
<h2></h2>
</div>
<div id="diffexp" class="section level2">
<h2>Differential gene expression</h2>
<p>Given an experiment with a least 1 design column with two values,
like wild-type (WT) vs knock out (of a specific gene), you can run
differential expression of genes. The output is an interactive plot,
where you can also search for you target genes, making it more useable
than normal expression plots, which often are very hard to read.</p>
<div id="selection-panel-differential-expression"
class="section level3">
<h3>Selection panel (Differential expression)</h3>
<p>Organism and experiment explained above - Differential method: FPKM
ratio is a pure FPKM ratio calculation without factor normalization
(like batch effects), fast and crude check. DESeq2 argument gives a
robust version, but only works for experiments with valid experimental
design (i.e. design matrix must be full ranked, see deseq2 tutorial for
details!) - Select two conditions (which 2 factors to group by)</p>
</div>
<div id="selection-panel-settings-2" class="section level3">
<h3>Selection panel (settings)</h3>
<ul>
<li>draw unregulated (show dots for unregulated genes, makes it much
slower!)</li>
<li>Full annotation (all transcript isoforms, default is primary isoform
only!)</li>
<li>P-value (sliding bar for p-value cutoff, default 0.05)</li>
<li>export format for plot (explained above)</li>
</ul>
</div>
</div>
<div id="read-length-qc" class="section level2">
<h2>Read length (QC)</h2>
<p>This tab displays a QC of pshifted coverage per readlength (like
start site of coding sequences) over all genes selected.</p>
<div id="display-panel-read-length-qc" class="section level3">
<h3>Display panel (Read length QC)</h3>
<p>The display panel shows what can be specified to display, the
possible select boxes are same as for heatmap above:</p>
</div>
<div id="plot-panel" class="section level3">
<h3>Plot panel</h3>
<p>From the options specified in the display panel, when you press
“plot” the data will be displayed. It contains the specific parts:</p>
<p>Top plot: Read length relative usage</p>
<ol style="list-style-type: decimal">
<li>Y-axis: Score</li>
<li>Color: Per frame (red, green, blue)</li>
<li>Facet box: the read length</li>
</ol>
<p>Bottom plot: Fourier transform (3 nucleotide periodicity quality,
clean peak means good periodicity)</p>
</div>
</div>
<div id="fastq-qc" class="section level2">
<h2>Fastq (QC)</h2>
<p>This tab displays the fastq QC output from fastp, as a html page.</p>
<div id="display-panel-read-length-qc-1" class="section level3">
<h3>Display panel (Read length QC)</h3>
<p>The display panel shows what can be specified to display, you can
select from organism, study and library.</p>
</div>
<div id="plot-panel-1" class="section level3">
<h3>Plot panel</h3>
<p>Displays the html page.</p>
</div>
</div>
</div>
<div id="metadata" class="section level1">
<h1>Metadata</h1>
<p>Metadata tab displays information about studies and custom
predictions. RiboCrypt is integrated with
<a href="https://rdp.ucc.ie/home" target="_blank">Ribo-seq Data
Portal</a>, refer to
<a href="https://doi.org/10.1093/nar/gkae1020" target="_blank">this
paper</a> for details on metadata curation and standardization.</p>
<p>## SRA search</p>
<p>Search SRA for full information of supported study</p>
<div id="study-accession-number" class="section level3">
<h3>Study accession number</h3>
<p>Here you input a study accession number in the form of either:</p>
<ul>
<li>SRP</li>
<li>GEO (GSE)</li>
<li>PRJNA (PRJ….)</li>
<li>PRJID (Only numbers)</li>
</ul>
</div>
<div id="output" class="section level3">
<h3>Output</h3>
<p>On top the abstract of the study is displayed, and on bottom a table
of all metadata found from the study is displayed.</p>
</div>
<div id="studies" class="section level2">
<h2>Studies</h2>
<p>Full table of supported studies with information about sample
counts</p>
</div>
<div id="translon" class="section level2">
<h2>Predicted translons</h2>
<p>Full list of predicted translons on all_merged tracks per species.
Each translons has a link that directs you to the browser tab and
displays the transcript zoomed in on the translons with a light yellow
highlighting of the region.</p>
<div id="settings-1" class="section level3">
<h3>Settings:</h3>
<p>Translon annotation scheme:</p>
<ul>
<li>Prediction algorithm: ORFik::detect_ribo_orfs</li>
<li>translon_types &lt;- c(“uORF”, “uoORF”,“doORF”,“dORF”)</li>
<li>start_codons &lt;- “ATG|TTG|CTG”</li>
<li>additional arguments: add sequence and amino acid sequence per
translon</li>
<li>All other arguments default.</li>
</ul>
</div>
</div>
</div>
<div id="additional-information" class="section level1">
<h1>Additional information</h1>
<p>All files are packed into ORFik experiments for easy access through
the ORFik backend package:</p>
<p>File formats used internally in experiments are:</p>
<ul>
<li>Annotation (gtf + TxDb for random access)</li>
<li>Fasta genome (.fasta, + index for random access)</li>
<li>Sequencing libraries (all duplicated reads are collapsed)
<ul>
<li>random access (only for collapsed read lengths): bigwig</li>
<li>Full genome coverage (only for collapsed read lengths): covRLE</li>
<li>Full genome coverage (split by read lengths): covRLElist</li>
</ul></li>
<li>count Tables (Summarized experiments, r data serialized .rds)</li>
<li>Library size list (Integer vector, .rds)</li>
<li>Precomputed gene coverages per organism: fst (used for
metabrowser)</li>
</ul>
<div id="massivengspipe" class="section level2">
<h2>massiveNGSpipe</h2>
<p>For our webpage the processing pipeline used is
<a href="https://github.com/rc-biotech/massiveNGSpipe" target="_blank">massiveNGSpipe</a>
which wraps multiple tools:</p>
<ol style="list-style-type: decimal">
<li>Fastq files are download with ORFik download.sra</li>
<li>Adapter is detected with either fastqc (sequence detection) and
falls back to fastp auto detection.</li>
<li>Reads are then trimmed with fastp (using the wrapper in ORFik)</li>
</ol>
<ul>
<li>Adapter removal specified</li>
<li>minimum read size (20nt)</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Read are collapsed (get the set of unique reads and put duplication
count in read header)</li>
<li>Reads are aligned with the STAR aligner (using the wrapper in
ORFik), that supports contamination removal. Settings:</li>
</ol>
<ul>
<li>genomic coordinates (to allow both genomic and transcriptomic
coordinates)</li>
<li>local alignment (to remove unknown flank effects)</li>
<li>minimum read size (20nt)</li>
</ul>
<ol start="6" style="list-style-type: decimal">
<li>When all samples of study are aligned, an ORFik experiment is
created that connects each sample to metadata (condition, inhibitor,
fraction, replicate etc)</li>
<li>Bam files are then converted to ORFik ofst format</li>
<li>These ofst files are then pshifted</li>
<li>Faster formats are then created (bigwig, fst and covRLE) for faster
visualization</li>
</ol>
</div>
<div id="api-for-url-access-and-sharing" class="section level2">
<h2>API for URL access and sharing</h2>
<p>RiboCrypt uses the shiny router API system for creating runable links
and backspacing etc. The API specificiation is the following:</p>
<div id="primary-url" class="section level3">
<h3>Primary url:</h3>
<p><a href="https://ribocrypt.org/"
class="uri">https://ribocrypt.org/</a> (This leads to browser page)</p>
</div>
<div id="page-selection-api" class="section level3">
<h3>Page selection API:</h3>
<p>Page selection is done with “#” followed by the page short name, the
list is the following:</p>
<ul>
<li>broser page (/ or /#browser)</li>
<li>MegaBrowser (/#MegaBrowser)</li>
<li>heatmap (/#heatmap)</li>
<li>codon (/#codon)</li>
<li>Differential expression (/#Differential expression)</li>
<li>Periodicity plot (/#periodicity)</li>
<li>fastq QC report (/#fastq)</li>
<li>SRA search (/#SRA search)</li>
<li>Studies supported (/#Studies)</li>
<li>Predicted Translons (/#Predicted Translons)</li>
<li>This tutorial (/#tutorial)</li>
</ul>
<p>Example: <a href="https://ribocrypt.org/#tutorial"
class="uri">https://ribocrypt.org/#tutorial</a> sends you to this
tutorial page</p>
</div>
<div id="API" class="section level3">
<h3>Parameter API:</h3>
<p>Settings can be specified by using the standard web parameter
API:</p>
<ul>
<li>“?”, Starts the parameter specification</li>
<li>“&amp;”, to combine terms:</li>
</ul>
<p>Example: <a
href="https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272#browser"
class="uri">https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272#browser</a>
will lead you to browser and insert gene ATF4 (all other settings being
default).</p>
<p>A more complicated call would be: <a
href="https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272&amp;tx=ENST00000404241&amp;frames_type=area&amp;kmer=9&amp;go=TRUE&amp;extendLeaders=100&amp;extendTrailers=100&amp;viewMode=TRUE&amp;other_tx=TRUE#browser"
class="uri">https://RiboCrypt.org/?dff=all_merged-Homo_sapiens&amp;gene=ATF4-ENSG00000128272&amp;tx=ENST00000404241&amp;frames_type=area&amp;kmer=9&amp;go=TRUE&amp;extendLeaders=100&amp;extendTrailers=100&amp;viewMode=TRUE&amp;other_tx=TRUE#browser</a></p>
<div id="browser-1" class="section level4">
<h4>browser:</h4>
<ul>
<li>dff=all_merged-Homo_sapiens (The Experiment to select: for webpage
it is “study id”_“Organism”)</li>
<li>gene=ATF4-ENSG00000128272 (For webpage it is: “Gene symbol”-“Ensembl
gene id”)</li>
<li>tx=ENST00000404241(isoform)</li>
<li>frames_type=area (plot type)</li>
<li>kmer=9 (window smoothing)</li>
<li>go=TRUE (plot on entry, you do not need to click plot for it to
happen)</li>
<li>extendLeaders=100 (extend 100 nt upstream of 5’ UTR / Leader)</li>
<li>extendTrailers=100 (extend 100 nt downstream of 3’ UTR /
Trailer)</li>
<li>viewMode=FALSE (If TRUE, Genomic coordinates, i.e. with
introns)</li>
<li>other_tx=TRUE (Full annotation, show transcript graph for all
genes/isoforms in area)</li>
<li>add_translon=FALSE (TRUE makes predicted translons be
displayed)</li>
<li>zoom_region: format tx coord: 20:50, format genomic coord:
1:2000-3000:p;1:4000:5000:p (p is converted to +, as url does not
support +)</li>
<li>#browser (open browser window)</li>
</ul>
</div>
</div>
</div>
<div id="about" class="section level2">
<h2>About</h2>
<p>This app is created as a collaboration with:</p>
<ul>
<li>University of Warsaw, Poland</li>
<li>University of Bergen, Norway</li>
</ul>
<p>Main authors and contact:</p>
<ul>
<li>Michal Swirski (Warsaw), email: <a
href="mailto:michal.swirski@uw.edu.pl"
class="email">michal.swirski@uw.edu.pl</a></li>
<li>Håkon Tjeldnes (Bergen), email: <a
href="mailto:hauken_heyken@hotmail.com"
class="email">hauken_heyken@hotmail.com</a></li>
</ul>
<script>
console.log("Tutorial JavaScript Loaded!"); // Debugging
</script>
<script>
window.addEventListener("message", function(event) {
  console.log("Received message:", event);  // Debugging
  if (event.origin !== window.location.origin) {
    console.log("Invalid origin:", event.origin);
    return;
  }

  if (event.data.scrollTo) {
    var target = document.getElementById(event.data.scrollTo);
    if (target) {
      console.log("Scrolling to:", event.data.scrollTo);  // Debugging
      target.scrollIntoView({ behavior: "smooth" });
    } else {
      console.log("Anchor not found:", event.data.scrollTo);  // Debugging
      console.log("Available elements in iframe:", document.body.innerHTML); // Debugging
    }
  }
});
</script>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
